Sources/NIO/Codec.swift

```diff
@@ -96,59 +96,83 @@ private extension ChannelHandlerContext {
 
 extension ByteToMessageDecoder {
 
+    /// Decode in a loop until there is nothing more to decode.
+    private func decodeLoop(ctx: ChannelHandlerContext, decodeFunc: (ChannelHandlerContext, inout ByteBuffer) throws -> DecodingState) {
+        assert(self.cumulationBuffer != nil)
+        ctx.withThrowingToFireErrorAndClose {
+            while true {
+                var buffer = self.cumulationBuffer!
+                let readable = buffer.readableBytes
+                let writerIndex = buffer.writerIndex
+                let result = try decodeFunc(ctx, &buffer)
+                if self.cumulationBuffer != nil {
+                    self.cumulationBuffer!.moveReaderIndex(forwardBy: readable - buffer.readableBytes)
+
+                    if result == .needMoreData {
+                        if self.cumulationBuffer!.writerIndex == writerIndex {
+                            // The user told use more data is needed and the cumulationBuffer did not change in between so we need to try again later.
+                            break
+                        }
+                    } else if self.cumulationBuffer!.readableBytes == 0 {
+                        // Nothing left to process, try again later.
+                        break
+                    }
+                } else {
+                    // The cumulationBuffer is nil which means we consumed everything via reentrance
+                    break
+                }
+            }
+        }
+    }
+
     /// Calls `decode` until there is nothing left to decode.
     public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-        var buffer = self.unwrapInboundIn(data)
         if self.cumulationBuffer != nil {
+            var buffer = self.unwrapInboundIn(data)
             self.cumulationBuffer!.write(buffer: &buffer)
-            buffer = self.cumulationBuffer!
         } else {
-            self.cumulationBuffer = buffer
+            self.cumulationBuffer = self.unwrapInboundIn(data)
         }
 
-        ctx.withThrowingToFireErrorAndClose {
-            // Running decode method until either the user returned `.needMoreData` or an error occurred.
-            while try decode(ctx: ctx, buffer: &buffer) == .`continue` && buffer.readableBytes > 0 { }
-        }
+        self.decodeLoop(ctx: ctx, decodeFunc: self.decode)
 
-        if buffer.readableBytes > 0 {
-            if self.shouldReclaimBytes(buffer: buffer) {
-                buffer.discardReadBytes()
+        if self.cumulationBuffer != nil {
+            if self.cumulationBuffer!.readableBytes > 0 {
+                if self.shouldReclaimBytes(buffer: self.cumulationBuffer!) && self.cumulationBuffer!.discardReadBytes() && self.cumulationBuffer!.readableBytes == 0 {
```

Comment:
Let’s make this clearer.

Reply:
Ok let me break up the if statement and add some comments